sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +376,368 @@ def transform(self, X):
+        If True, a copy of X will be created. If False, imputation will
+        be done in-place whenever possible. Note that, if metric is
+        "masked_euclidean" and copy=False then missing_values in the
+        input matrix X will be overwritten with zeros.
+
+    Attributes
+    ----------
+    statistics_ : 1-D array of length {n_features}
+        The 1-D array contains the mean of each feature calculated using
+        observed (i.e. non-missing) values. This is used for imputing
+        missing values in samples that are either excluded from nearest
+        neighbors search because they have too many ( > row_max_missing)
+        missing features or because all of the sample's k-nearest neighbors
+        (i.e., the potential donors) also have the relevant feature value
+        missing.
+
+    References
+    ----------
+    * Olga Troyanskaya, Michael Cantor, Gavin Sherlock, Pat Brown, Trevor
+      Hastie, Robert Tibshirani, David Botstein and Russ B. Altman, Missing
+      value estimation methods for DNA microarrays, BIOINFORMATICS Vol. 17
+      no. 6, 2001 Pages 520-525.
+
+    Examples
+    --------
+    >>> from sklearn.preprocessing.imputation import KNNImputer
+    >>> nan = float("NaN")
+    >>> X = [[1, 2, nan], [3, 4, 3], [nan, 6, 5], [8, 8, 7]]
+    >>> imputer = KNNImputer(max_neighbors=2, weights="uniform")
+    >>> imputer.fit_transform(X)
+    array([[ 1. ,  2. ,  4. ],
+           [ 3. ,  4. ,  3. ],
+           [ 5.5,  6. ,  5. ],
+           [ 8. ,  8. ,  7. ]])
+    """
+
+    def __init__(self, missing_values="NaN", max_neighbors=5,
+                 weights="uniform", metric="masked_euclidean",
+                 row_max_missing=0.5, col_max_missing=0.8, copy=True):
+
+        self.missing_values = missing_values
+        self.max_neighbors = max_neighbors
+        self.weights = weights
+        self.metric = metric
+        self.row_max_missing = row_max_missing
+        self.col_max_missing = col_max_missing
+        self.copy = copy
+
+    def fit(self, X, y=None):
+        """Fit the imputer on X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape (n_samples, n_features)
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        # Imports here to avoid circular import
+        from ..neighbors import NearestNeighbors
+        from ..neighbors.base import _check_weights
+
+        # Check parameters
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=np.float64,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        self.weights = _check_weights(self.weights)
+
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
```

Comment:
don't we check that in the metric, or do we get issues with ``_get_mask`` otherwise?